File: C++/DesignPattern/CreationalDesignPattern/SingletonDesignPattern.cpp
// Singleton Design Pattern
#include <iostream>
#include <vector>
#include <string>
#include <mutex>
#include <thread>
using namespace std;

class Logger {
    static Logger* instance;
    static mutex mtx;

    Logger() {
        cout << "object created\n";
    }

public:
    vector<string> logs;

    // Thread-unsafe version
    static Logger* getInstanceThreadUnsafe() {
        if (instance == nullptr) {
            instance = new Logger();
        }
        return instance;
    }

    // Thread-safe: lock on every call
    static Logger* getInstanceSynchronized() {
        lock_guard<mutex> lock(mtx);
        if (instance == nullptr) {
            instance = new Logger();
        }
        return instance;
    }

    // Thread-safe: double-checked locking
    static Logger* getInstance() {
        if (instance == nullptr) {
            lock_guard<mutex> lock(mtx);
            if (instance == nullptr) {
                instance = new Logger();
            }
        }
        return instance;
    }

    void appendLogs(const string& log) {
        logs.push_back(log);
    }
};

Logger* Logger::instance = nullptr;
mutex Logger::mtx;

int main() {
    auto task = []() {
        Logger::getInstance();
    };

    vector<thread> threads;
    for (int i = 0; i < 2; i++) {
        threads.emplace_back(task);
    }
    for (auto& t : threads) {
        t.join();
    }

    return 0;
}

File: Python/DesignPattern/CreationalDesignPattern/SingletonDesignPattern.py
# Singleton Design Pattern
import threading


class Logger:
    _instance = None
    _lock = threading.Lock()

    def __init__(self):
        print("object created")
        self.logs = []

    def append_logs(self, log: str):
        self.logs.append(log)

    # Thread-unsafe version
    @staticmethod
    def get_instance_thread_unsafe():
        if Logger._instance is None:
            Logger._instance = Logger()
        return Logger._instance

    # Thread-safe: synchronized (lock on every call)
    @staticmethod
    def get_instance_synchronized():
        with Logger._lock:
            if Logger._instance is None:
                Logger._instance = Logger()
        return Logger._instance

    # Thread-safe: double-checked locking
    @staticmethod
    def get_instance():
        if Logger._instance is None:
            with Logger._lock:
                if Logger._instance is None:
                    Logger._instance = Logger()
        return Logger._instance


if __name__ == "__main__":
    def task():
        Logger.get_instance()

    threads = [threading.Thread(target=task) for _ in range(2)]
    for t in threads:
        t.start()
    for t in threads:
        t.join()
